constraints_lang.md

```diff
@@ -0,0 +1,311 @@
+
+Example 1
+
+Say that a user has a path A -> B -> C and wants to make ||AB|| be derived from ||BC||. The user must rewrite their code so that the path is constructed in a different way. A few options:
+
+1. Reverse the order: C -> B -> A -- Requires rewriting delta expressions, e.g. angles are in the opposite direction.
+2. Rotate the path: B -> C -> A -> B -- Only works for closed paths. It moves the root, but doesn't remove it.
+3. Factor commonality out: `let x = common(); AB = segment1(x); BC = segment2(x)` -- `x` and `common` are likely very specific and can't be used anywhere else. `x` needs a name in the source, and it may not have much meaning on its own. Segment construction functions (`segment1` and `segment2` in this example) need to allow `x` to be an input. In the worst case, the implementation of `segment1()` might need to be: `if bar(x) { ... } else { ... }`. If we want to allow all this to happen with point-and-click, the app needs to understand how to write `common`, `segment1`, and `segment2` in terms of the original path expression. This seems non-trivial in the general case because anything that can be computed in stdlib functions must have a pure counterpart that can be used in `common`. `common` would ideally not have the side-effect of drawing to reduce ordering restrictions.
+
+Example 2
+
+Say that a user wants to construct a quadrilateral with ordered points A, B, C, and D such that ||CD|| is derived from ||AB|| and ||BC|| is derived from ||DA||. There are dependencies in both directions. Workarounds (1) and (2) above do not work since no single topological sort exists. The user _must_ factor out common expressions.
+
+Factoring out code seems like a workaround for a fundamental limitation of KCL, relative to existing CAD software. Even when constraints are possible in KCL today, it requires that users write them down in a very specific way that isn't necessarily intuitive, even using non-visual linguistic mathematical reasoning.
+# Proposal: A Constraint Language
+
+A constraint in the mechanical engineering world is a mathematical relationship between two expressions. The user should be able to write these relationships in any order. The language implementation should treat these relationships as a system of constraints and solve for whatever derived values that are required to construct geometry. Derived values should not need to be expressed as the outputs of functions with known inputs if their relationship is determined by an equation that can be symbolically rearranged to the same thing.
+
+As an example, we'll try to construct a sketch of a polygon where dependencies flow in both directions.
+
+Note that the code is generated by the user clicking in the UI. The resulting code is in the order that it was created, meaning it's a history of actions that the user performed in the UI.
+
+```
+p1 = screen_point(x: 10, y: 12)
+p2 = screen_point(x: 20, y: 13)
+l1 = line(p1, p2)
```

Comment:
During KittyCAMP 2024 we discussed the benefits of "having every segment be relative unless a 'fixed' constraint is applied". How are relative lines represented in this approach at all?

Reply:
Yes, you're absolutely right. I completely screwed this up. I want everything to be relative. I will need to change a few things. Maybe this will inform how to fix `screen_point()` et al.

Reply:
Sick, okay that makes sense though. Maybe the point definitions continue to be "absolute" (your adjustments to `screen_point()` notwithstanding), and it's actually the _constraints that drive their positions_, not the other way round. I'm not sure. I know that's why you can sometimes get unexpected movement of points on mostly-constrained sketches in solver-based CAD programs like Onshape; it's a kind of drift that occurs due to this relative/absolute tension. I can't seem to create an example video for you but I will post it here if I manage it.